### PR TITLE
E2E llm chat template tests

### PR DESCRIPTION
**Pull request by rasapala**
### 🛠 Summary

JIRA CVS-144457
More chat template tests.

### 🧪 Checklist

- [x] Unit tests added.
- [x] The documentation updated.
- [x] Change follows security best practices.
``

### CODE REVIEW

File: src/test/llmtemplate_test.cpp [hunk 1]
```diff
@@ -38,6 +42,28 @@
 
 using namespace ovms;
 
+class MockedServerRequestInterface final : public tensorflow::serving::net_http::ServerRequestInterface {
```
Comment (dkalinowski) [inline src/test/llmtemplate_test.cpp:45]: Already exists in two places - llmnode_test and http_openai_handler_test, and now we introduce into third test case, maybe we could separate to test utils?

File: src/test/llmtemplate_test.cpp [hunk 2]
```diff
@@ -348,3 +377,291 @@ TEST_F(LLMChatTemplateTest, ChatTemplateTwoConfigs) {
...
+            output_stream: "LOOPBACK:loopback"
+            output_stream: "HTTP_RESPONSE_PAYLOAD:output"
+            input_stream_info: {
+            tag_index: 'LOOPBACK:0',
+            back_edge: true
+            }
+            node_options: {
+                [type.googleapis.com/mediapipe.LLMCalculatorOptions]: {
+                models_path: "MODELS_PATTERN",
+                plugin_config: "{\"INFERENCE_PRECISION_HINT\":\"f32\"}"
```
Comment (dkalinowski) [inline src/test/llmtemplate_test.cpp:411]: In my tests llama3 segfaults on spr-36 when setting plugin config f32

File: src/test/llmtemplate_test.cpp [hunk 3]
```diff
@@ -348,3 +377,291 @@ TEST_F(LLMChatTemplateTest, ChatTemplateTwoConfigs) {
...
+
+    void CreateSymbolicLinks() {
+        for (const auto& entry : fs::directory_iterator(ONE_MODEL_PATH)) {
+            std::filesystem::path outfilename = entry.path();
+            std::string outfilename_str = outfilename.string();
+            std::string fileName = GetFileNameFromPath(ONE_MODEL_PATH, outfilename_str);
+            SPDLOG_INFO("Filename to link {}\n", fileName);
+            std::string symlinkPath = ovms::FileSystem::joinPath({directoryPath, fileName});
+            SPDLOG_INFO("Creating symlink from: {}\n to:\n{}", outfilename_str, symlinkPath);
+            fs::create_symlink(outfilename_str, symlinkPath);
```
Comment (dkalinowski) [inline src/test/llmtemplate_test.cpp:465]: Can we create guard classes for the symlinks? This way we are sure that every symlink created in unit test will be destroyed upon successful test run **and also** fail at any point.

File: src/test/llmtemplate_test.cpp [hunk 4]
```diff
@@ -348,3 +377,291 @@ TEST_F(LLMChatTemplateTest, ChatTemplateTwoConfigs) {
...
+            "model": "llmDummyKFS",
+            "stream": true,
+            "seed" : 1,
+            "max_tokens": 5,
+            "prompt": "?"
+        }
+    )";
+
+    EXPECT_CALL(writer, PartialReplyEnd()).Times(1);
+    EXPECT_CALL(writer, PartialReply(::testing::_)).Times(5);
```
[the remark this answers is not in the document]
Comment (rasapala) [reply src/test/llmtemplate_test.cpp:647]: Content is checked below.

File: src/test/llmtemplate_test.cpp [hunk 5]
```diff
@@ -348,3 +377,291 @@ TEST_F(LLMChatTemplateTest, ChatTemplateTwoConfigs) {
...
+            }
+            ]
+        }
+    )";
+
+    ASSERT_EQ(
+        handler->dispatchToProcessor(endpointChatCompletions, requestBody, &response, comp, responseComponents, &writer),
+        ovms::StatusCode::OK);
+    // Assertion split in two parts to avoid timestamp missmatch
+    const size_t timestampLength = 10;
```
[the remark this answers is not in the document]
Comment (rasapala) [reply src/test/llmtemplate_test.cpp:544]: Already used in other stream tests.

File: src/test/llmtemplate_test.cpp [hunk 6]
```diff
@@ -348,3 +355,299 @@ TEST_F(LLMChatTemplateTest, ChatTemplateTwoConfigs) {
...
+        ovmsServer.shutdownModules();
+    }
+};
+
+class LLMChatTemplateHttpTest : public TestWithTempDir {
+protected:
+    static std::unique_ptr<std::thread> t;
+
+    const std::string GRAPH_PATTERN = "GRAPH_PATTERN";
+    const std::string WORKSPACE_PATTERN = "MODELS_PATTERN";
```
Comment (mzegla) [inline src/test/llmtemplate_test.cpp:419]:
```suggestion
    const std::string GRAPH_PATTERN = "<GRAPH_PATTERN>";
    const std::string WORKSPACE_PATTERN = "<MODELS_PATTERN>";
```
In other places we use put parts that we want to substitute later in `<>`

File: src/test/llmtemplate_test.cpp [hunk 7]
```diff
@@ -348,3 +355,299 @@ TEST_F(LLMChatTemplateTest, ChatTemplateTwoConfigs) {
...
+    const std::string ONE_MODEL_PATH = "/ovms/llm_testing/facebook/opt-125m";
```
Comment (mzegla) [inline src/test/llmtemplate_test.cpp:420]:
```suggestion
    const std::string MODEL_PATH = "/ovms/llm_testing/facebook/opt-125m";
```

File: src/test/llmtemplate_test.cpp [hunk 8]
```diff
@@ -348,3 +355,299 @@ TEST_F(LLMChatTemplateTest, ChatTemplateTwoConfigs) {
...
+    bool CreatePipelineGraph(const std::string& workspacePath) {
+        std::string configContents = graphTemplate;
+        configContents.replace(configContents.find(WORKSPACE_PATTERN), std::string(WORKSPACE_PATTERN).size(), workspacePath);
+        return CreateConfig(configContents, graphConfigFilePath);
+    }
+
+    void CreateSymbolicLinks() {
+        for (const auto& entry : fs::directory_iterator(ONE_MODEL_PATH)) {
+            std::filesystem::path outfilename = entry.path();
+            std::string outfilename_str = outfilename.string();
```
Comment (mzegla) [inline src/test/llmtemplate_test.cpp:448]: Please follow camelCase

File: src/test/llmtemplate_test.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (atobiszei) [reply src/test/llmtemplate_test.cpp:654]: Where? I see checking in L609-612 but not here and inferChatCompletionsStream.

File: src/test/llmtemplate_test.cpp [hunk 9]
```diff
@@ -348,3 +355,306 @@ TEST_F(LLMChatTemplateTest, ChatTemplateTwoConfigs) {
...
+
+    ASSERT_EQ(
+        handler->dispatchToProcessor(endpointCompletions, requestBody, &response, comp, responseComponents, &writer),
+        ovms::StatusCode::OK);
+    // Assertion split in two parts to avoid timestamp missmatch
+    const size_t timestampLength = 10;
+    std::string expectedResponsePart1 = R"({"choices":[{"finish_reason":"stop","index":0,"logprobs":null,"text":"\n\nThe first thing"}],"created":)";
+    std::string expectedResponsePart2 = R"(,"model":"llmDummyKFS","object":"text_completion"})";
+    ASSERT_EQ(response.compare(0, expectedResponsePart1.length(), expectedResponsePart1), 0);
+    ASSERT_EQ(response.compare(expectedResponsePart1.length() + timestampLength, expectedResponsePart2.length(), expectedResponsePart2), 0);
```
Comment (mzegla) [inline src/test/llmtemplate_test.cpp:612]: Is it deterministic that we will get exactly this output and in this order?

**Review by atobiszei — APPROVED** [review]

**Review by mzegla — APPROVED** [review]

**Pull request merged**